File: swift/cell.swift
final public class Cell {
  public let x: UInt32
  public let y: UInt32
  public var alive: Bool
  public var nextState: Bool?
  public var neighbours: [Cell]

  public init(x: UInt32, y: UInt32, alive: Bool = false) {
    self.x = x
    self.y = y
    self.alive = alive
    self.nextState = nil
    self.neighbours = []
  }

  public func toChar() -> String {
    return alive ? "o" : " "
  }

  public func aliveNeighbours() -> UInt32 {
    // The following is the fastest
    return UInt32(neighbours.count(where: \.alive))

    // The following is slower
    // var aliveNeighbours = UInt32(0);
    // for neighbour in neighbours {
    //   if neighbour.alive {
    //     aliveNeighbours += 1
    //   }
    // }
    // return aliveNeighbours

    // The following is slower
    // var aliveNeighbours = UInt32(0)
    // let count = neighbours.count
    // for i in 0 ..< count {
    //   let neighbour = neighbours[i]
    //   if neighbour.alive {
    //     aliveNeighbours += 1
    //   }
    // }
    // return aliveNeighbours
  }
}

File: nim/cell.nim
import std/options
# from std/sequtils import filter

type
  Cell = ref object
    x: uint32
    y: uint32
    alive: bool
    nextState: Option[bool]
    neighbours: seq[Cell]

func toChar(self: Cell): string =
  if self.alive:
    "o"
  else:
    " "

func aliveNeighbours(self: Cell): uint32 =
  # The following is slower
  # uint32(
  #   filter(
  #     self.neighbours,
  #     proc(cell: Cell): bool = cell.alive
  #   ).len
  # )

  # The following is slower
  # var aliveNeighbours = 0'u32
  # for neighbour in self.neighbours:
  #   if neighbour.alive:
  #     aliveNeighbours += 1
  # aliveNeighbours

  # The following is the fastest
  var aliveNeighbours = 0'u32
  let count = self.neighbours.len
  for i in 0..<count:
    let neighbour = self.neighbours[i]
    if neighbour.alive:
      aliveNeighbours += 1
  aliveNeighbours
